MFA token config flow

shtrom opened the pull request:
Based on #9, fixes #10.

* add a ConfigFlow module, so the token can be renewed in the UI, along with nice notifications when it gets outdated
* bump version to v2.0.0 due to breaking change in the entity names (addition of the service_agreement_id to support multiple accounts, but needing re-adding to the energy dashboard), and the change in configuration method.
* support AuroraPlus's MFA authentication

This PR requires an AuroraPlus Python module with 2FA support, from https://github.com/LeighCurran/AuroraPlus/pull/5. If you have a previously setup AuroraPlusHA, you may encounter issues due to using the old version of the module. You'll have to remove the installed files, and completely restart your Home Assistant system so it installs the new version. In Dockerised setups, this can be done with a variation of ```docker compose exec homeassistant rm -rf /usr/local/lib/python3.13/site-packages/auroraplus\*```

With HAOS, you may need [Advanced SSH & Web Terminal](https://github.com/hassio-addons/addon-ssh) with disabled protection mode to be able to run that command. Please think through this and make sure you understand what you are doing. While this particular suggestion is not particularly risky (says I), you shouldn't just trust strangers on the Internet without verifying.

shtrom committed:
autopep linting

Signed-off-by: Olivier Mehani <shtrom@ssji.net>

shtrom committed:
Switch kWh usage to STATE_CLASS_TOTAL_INCREASING to avoid reset artefacts

fixes: #6

Signed-off-by: Olivier Mehani <shtrom@ssji.net>

shtrom committed:
Add per-tariff summary

Tariffs 31, 41, 61 and 62 are a wild guess, and totally untested.

fixes: #6 some more
Signed-off-by: Olivier Mehani <shtrom@ssji.net>

shtrom committed:
Use STATE_CLASS_TOTAL with last_reset for everything

Neither STATE_CLASS_TOTAL_INCREASING nor STATE_CLASS_MEASUREMENT work
well for the ca. 9am reset that seems to happen for daily summary data.

The `last_reset` time is set to whenever a value changes. This means
this is liable to a bug where a value will be missed it it is the same
as the previous day.

Signed-off-by: Olivier Mehani <shtrom@ssji.net>

shtrom committed:
Make update asynchronous

Signed-off-by: Olivier Mehani <shtrom@ssji.net>

shtrom committed:
Split out AuroraPlus API coordinator

Signed-off-by: Olivier Mehani <shtrom@ssji.net>

shtrom committed:
Sanitise entity_id

Signed-off-by: Olivier Mehani <shtrom@ssji.net>

shtrom committed:
Add HistoricalSensor for daily per-tariff statistics #6

Signed-off-by: Olivier Mehani <shtrom@ssji.net>

shtrom committed:
Fix Throttle behaviour

The AuroraApi proxies everything to the session, or swallows the error
transparently.  This is not the behaviour that the Throttle wrapper
expects, as it needs to use a _throttle attribute on the wrapped object.

We add a special case in the proxy juuuuust for this.

Signed-off-by: Olivier Mehani <shtrom@ssji.net>

shtrom committed:
Cleanup logging

Signed-off-by: Olivier Mehani <shtrom@ssji.net>

shtrom committed:
Be a bit more defensive on empty data

Signed-off-by: Olivier Mehani <shtrom@ssji.net>

shtrom committed:
Update to auroraplus 1.5.0

Signed-off-by: Olivier Mehani <shtrom@ssji.net>

shtrom committed:
Only create entities for available tariffs

Signed-off-by: Olivier Mehani <shtrom@ssji.net>

shtrom committed:
Look backwards in time for valid data when updating

Signed-off-by: Olivier Mehani <shtrom@ssji.net>

shtrom committed:
[REVERTME] Point to AuroraPLus fork

Waiting for LeighCurran/AuroraPlus#1 to be merged.

Signed-off-by: Olivier Mehani <shtrom@ssji.net>

shtrom committed:
Get Tariffs from monthly summary

Signed-off-by: Olivier Mehani <shtrom@ssji.net>

shtrom committed:
Add per-tariff daily cost sensors

Signed-off-by: Olivier Mehani <shtrom@ssji.net>

shtrom committed:
Try to improve startup reliability and performance

Signed-off-by: Olivier Mehani <shtrom@ssji.net>

shtrom committed:
Saner log verbosity

Signed-off-by: Olivier Mehani <shtrom@ssji.net>

shtrom committed:
Don't round values in stats

Signed-off-by: Olivier Mehani <shtrom@ssji.net>

shtrom committed:
Use absolute values on hourly measurements

This is useful for the T140 feed-in, which is negative, when HA expect a
positive return value.

Signed-off-by: Olivier Mehani <shtrom@ssji.net>

shtrom committed:
Update AuroraPlus branch for MFA

Usage:

1. obtain an `access_token` following instructions at [0]
2. update the secrets to use an empty username (`''`), and the
   `access_token` as the password

[0] https://github.com/shtrom/AuroraPlus/tree/oauth-mfa-token\#obtain-a-token

Signed-off-by: Olivier Mehani <shtrom@ssji.net>

shtrom committed:
Bump homeassistant-historical-sensor to 2.0.0rc5

Fixes

    Traceback (most recent call last):
    File /usr/src/homeassistant/homeassistant/loader.py, line 842, in get_platform
    cache[full_name] = self._import_platform(platform_name)
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    File /usr/src/homeassistant/homeassistant/loader.py, line 859, in _import_platform
    return importlib.import_module(f{self.pkg_path}.{platform_name})
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    File /usr/local/lib/python3.11/importlib/init.py, line 126, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    File , line 1204, in _gcd_import
    File , line 1176, in _find_and_load
    File , line 1147, in _find_and_load_unlocked
    File , line 690, in _load_unlocked
    File , line 940, in exec_module
    File , line 241, in _call_with_frames_removed
    File /config/custom_components/auroraplus/sensor.py, line 39, in
    from homeassistant_historical_sensor import (
    File /usr/local/lib/python3.11/site-packages/homeassistant_historical_sensor/init.py, line 39, in
    raise SystemError(msg)
    SystemError: Running HomeAssistant 2024.1.2, Minimum required version >=2023.6.0

See [0], [1]

[0] ldotlopez/ha-historical-sensor#12
[1] https://github.com/LeighCurran/AuroraPlus/pull/3\#issuecomment-1884213483

Signed-off-by: Olivier Mehani <shtrom@ssji.net>

shtrom committed:
Add config_flow

Nicer entities names

Update README for Config Flow

Move AuroraApi to... api

Signed-off-by: Olivier Mehani <shtrom@ssji.net>

shtrom committed:
Bump version to 2.0.0 due to breaking changes

* Config is not migrated
* entity_ids now contain the service_agreement_id, and need to be
  re-added to the energy dashboard configuration

Signed-off-by: Olivier Mehani <shtrom@ssji.net>

shtrom commented:
@LeighCurran  I thought I had this PR around before, but can't find it. I now added a ConfigFlow module, so the token can be renewed in the UI, along with nice notifications when it gets outdated. It's updating to v2.0.0 due to breaking change in the entity names (needing re-adding to the energy dashboard), and the change in configuration method (for which I haven't added a migration step).

@slothking87 wanna give it a try, if you're game?

shtrom commented:
There is a persistent, but seemingly harmless, bug that occurs after updating the access_token, and continues until HA is restarted.

```
homeassistant  | 2024-07-03 12:03:49.098 ERROR (Recorder) [homeassistant.components.recorder.core] SQLAlchemyError error processing task CommitTask()                   
homeassistant  | Traceback (most recent call last):                                                                                                                     
homeassistant  |   File "/usr/src/homeassistant/homeassistant/components/recorder/core.py", line 926, in _process_one_task_or_event_or_recover                          
homeassistant  |     task.run(self)                                                                                                                                     
homeassistant  |   File "/usr/src/homeassistant/homeassistant/components/recorder/tasks.py", line 291, in run                                                           
homeassistant  |     instance._commit_event_session_or_retry()  # noqa: SLF001                                                                                          
homeassistant  |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^                                                                                                          
homeassistant  |   File "/usr/src/homeassistant/homeassistant/components/recorder/core.py", line 1191, in _commit_event_session_or_retry                                
homeassistant  |     self._commit_event_session()                                                                                                                       
homeassistant  |   File "/usr/src/homeassistant/homeassistant/components/recorder/core.py", line 1217, in _commit_event_session                                         
homeassistant  |     session.execute(                                                                                                                                   
homeassistant  |   File "/usr/local/lib/python3.12/site-packages/sqlalchemy/orm/session.py", line 2351, in execute                                                      
homeassistant  |     return self._execute_internal(                                                                                                                     
homeassistant  |            ^^^^^^^^^^^^^^^^^^^^^^^
homeassistant  |   File "/usr/local/lib/python3.12/site-packages/sqlalchemy/orm/session.py", line 2236, in _execute_internal
homeassistant  |     result: Result[Any] = compile_state_cls.orm_execute_statement(
homeassistant  |                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
homeassistant  |   File "/usr/local/lib/python3.12/site-packages/sqlalchemy/orm/bulk_persistence.py", line 1603, in orm_execute_statement
homeassistant  |     result = _bulk_update(
homeassistant  |              ^^^^^^^^^^^^^
homeassistant  |   File "/usr/local/lib/python3.12/site-packages/sqlalchemy/orm/bulk_persistence.py", line 335, in _bulk_update
homeassistant  |     persistence._emit_update_statements(
homeassistant  |   File "/usr/local/lib/python3.12/site-packages/sqlalchemy/orm/persistence.py", line 948, in _emit_update_statements
homeassistant  |     raise orm_exc.StaleDataError(
homeassistant  | sqlalchemy.orm.exc.StaleDataError: UPDATE statement on table 'states' expected to update 1 row(s); 0 were matched.
```

I haven't gotten to the bottom of it yet, but I suspect this has to do with using the historical sensor, and HA trying to update the current (non-existent) reading. @ldotlopez, perhaps you have seen a similar issue before?

ldotlopez commented:
> I haven't gotten to the bottom of it yet, but I suspect this has to do with using the historical sensor, and HA trying to update the current (non-existent) reading. @ldotlopez, perhaps you have seen a similar issue before (the issue also happens with 2.0.0rc6)?

Yes, I have this issue too. 

For me, It surfaces when two historical sensors are enabled for the same integration. I spent so many hours in this issue that gives me headaches to think about it.

Historical-sensors consists of two main parts: handling historical states and handling (import) historical statistics.

For importing statistics H-A has an API, but it doesn't for the first one. For this reason historical-sensors manipulates the DB "directly" (using SQLalchemy and some helper functions from the recorder component). The problem is that H-A recently added some layer on top of the DB that I was unable to made work with historical states.

The only solution I can see now is to remove the handling of historical states and just transform upstream data into statistics.

shtrom commented:
Yep, cool, @ldotlopez. Thanks for the details!

What do you mean by “remove the handling of historical states and just transform upstream data into statistics”? This sounds like dropping the use of ha-historical-sensors? Or is this just something like disabling stats, but not raw data input? I remember struggling to get stats working, but can't recall the details.

In any case, we can't do that here, as we never get immediate data, just full-day hourly data, the next day.

It's ok, though, with a restart, the problem clears and everything works, so I'll look at a way to make the config update make HA ask for a reboot.

slothking87 commented:
> @LeighCurran I thought I had this PR around before, but can't find it. I now added a ConfigFlow module, so the token can be renewed in the UI, along with nice notifications when it gets outdated. It's updating to v2.0.0 due to breaking change in the entity names (needing re-adding to the energy dashboard), and the change in configuration method (for which I haven't added a migration step).
> 
> @slothking87 wanna give it a try, if you're game?

Sorry - Been away and forgot to have a look.
Works perfectly as far as I can tell so far!

shtrom committed:
bump homeassistant-historical-sensor to 2.0.0rc6

shtrom committed:
new version 2.0.1

shtrom committed:
Use hass.config_entries.async_forward_entry_setups

This is required from HA 2025.06 [0].

[0] https://developers.home-assistant.io/blog/2024/06/12/async_forward_entry_setups/

shtrom committed:
Add link to Access Token-fetching page

shtrom committed:
Raise ConfigEntryAuthFailed on authorisation errors when fetching data

slothking87 commented:
@shtrom  - my HA keeps needing authentication daily at the moment. Are you seeing the same behaviour?

shtrom commented:
Yep. The access token lifetime seems to have been shortened. 

Fortunately, there is a non-standard (AFAIK) way to refresh the token.

I have a fix in the work.
-- 
Olivier Mehani ***@***.***>
Sent from my mobile, please excuse my brevity.

shtrom committed:
Use id_token to refresh access_token

shtrom committed:
[REVERTME] Point to AuroraPlus branch with refresh support

shtrom commented:
Ok, here we go. I pushed a few changes to use the `id_token` to get what's needed to refresh the `access_token`. This requires some new work from the AuroraPlus lib https://github.com/LeighCurran/AuroraPlus/pull/5, so may have to do the same dance as usual for the update.

Also, I created a small page to allow getting the tokens from the browser: https://shtrom.github.io/AuroraPlus/

BEWARE: it fetches tokens that allow to manipulate your AuroraPlus account. It runs entirely in your browser, and nothing should leak out (at least not to me). The code is here https://github.com/shtrom/AuroraPlus/tree/pages

shtrom committed:
manifest: add loggers

slothking87 commented:
This is highly likely due to the python script not updating properly (again)
- Im getting this error - Ill have another go at it tomorrow to try make it take hold. 

Logger: aiohttp.server
Source: /usr/local/lib/python3.13/site-packages/aiohttp/web_protocol.py:481
First occurred: 9:25:40 PM (10 occurrences)
Last logged: 9:58:00 PM

Error handling request from 192.168.68.92
Traceback (most recent call last):
  File "/usr/local/lib/python3.13/site-packages/aiohttp/web_protocol.py", line 510, in _handle_request
    resp = await request_handler(request)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.13/site-packages/aiohttp/web_app.py", line 569, in _handle
    return await handler(request)
           ^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.13/site-packages/aiohttp/web_middlewares.py", line 117, in impl
    return await handler(request)
           ^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/components/http/security_filter.py", line 92, in security_filter_middleware
    return await handler(request)
           ^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/components/http/forwarded.py", line 83, in forwarded_middleware
    return await handler(request)
           ^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/components/http/request_context.py", line 26, in request_context_middleware
    return await handler(request)
           ^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/components/http/ban.py", line 86, in ban_middleware
    return await handler(request)
           ^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/components/http/auth.py", line 242, in auth_middleware
    return await handler(request)
           ^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/components/http/headers.py", line 41, in headers_middleware
    response = await handler(request)
               ^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/helpers/http.py", line 73, in handle
    result = await handler(request, **request.match_info)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/components/http/decorators.py", line 83, in with_admin
    return await func(self, request, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/components/config/config_entries.py", line 227, in post
    return await super().post(request, flow_id)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/components/http/data_validator.py", line 74, in wrapper
    return await method(view, request, data, *args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/helpers/data_entry_flow.py", line 121, in post
    result = await self._flow_mgr.async_configure(flow_id, data)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/data_entry_flow.py", line 336, in async_configure
    result = await self._async_configure(flow_id, user_input)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/src/homeassistant/homeassistant/data_entry_flow.py", line 383, in _async_configure
    result = await self._async_handle_step(
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
        flow, cur_step["step_id"], user_input
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    )
    ^
  File "/usr/src/homeassistant/homeassistant/data_entry_flow.py", line 487, in _async_handle_step
    result: _FlowResultT = await getattr(flow, method)(user_input)
                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/config/custom_components/auroraplus/config_flow.py", line 52, in async_step_user
    return await self._configure(user_input)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/config/custom_components/auroraplus/config_flow.py", line 65, in _configure
    session = await self.hass.async_add_executor_job(aurora_init, id_token)
              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.13/concurrent/futures/thread.py", line 59, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/config/custom_components/auroraplus/api.py", line 25, in aurora_init
    session = auroraplus.api(id_token=id_token)
TypeError: api.__init__() got an unexpected keyword argument 'id_token'

shtrom commented:
On 10 August 2025 10:03:43 pm AEST, slothking87 ***@***.***> wrote:
>  File "/usr/local/lib/python3.13/concurrent/futures/thread.py", line 59, in run
>    result = self.fn(*self.args, **self.kwargs)
>  File "/config/custom_components/auroraplus/api.py", line 25, in aurora_init
>    session = auroraplus.api(id_token=id_token)
>TypeError: api.__init__() got an unexpected keyword argument 'id_token'
>

Yep. I had to go in an delete the old `auroraplus` from site-packages in the homeassistant container in HAOS:
-- 
Olivier Mehani ***@***.***>
Sent from my mobile, please excuse my brevity.

slothking87 commented:
At the risk of asking a dumb question - Which integration AuroraPlusHA
version should I be using with which python script? I have a feeling Im
trying to use the wrong branch


On Mon, Aug 11, 2025 at 8:12 AM Olivier Mehani ***@***.***>
wrote:

> *shtrom* left a comment (LeighCurran/AuroraPlusHA#11)
> <https://github.com/LeighCurran/AuroraPlusHA/pull/11#issuecomment-3172927267>
> On 10 August 2025 10:03:43 pm AEST, slothking87 ***@***.***> wrote:
> > File "/usr/local/lib/python3.13/concurrent/futures/thread.py", line 59,
> in run
> > result = self.fn(*self.args, **self.kwargs)
> > File "/config/custom_components/auroraplus/api.py", line 25, in
> aurora_init
> > session = auroraplus.api(id_token=id_token)
> >TypeError: api.__init__() got an unexpected keyword argument 'id_token'
> >
>
> Yep. I had to go in an delete the old `auroraplus` from site-packages in
> the homeassistant container in HAOS:
> --
> Olivier Mehani ***@***.***>
> Sent from my mobile, please excuse my brevity.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/LeighCurran/AuroraPlusHA/pull/11#issuecomment-3172927267>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AM7YHPRWRSTWPABFR6XB3RT3M67WVAVCNFSM6AAAAAB72BIVU2VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTCNZSHEZDOMRWG4>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

shtrom commented:
> Somewhere something doesnt like me. I have nuked both the integration and the deps folder a number of times tonight, have made sure that Im downloading the correct one, but still get errors like this when i try to save the token against the integration. Ill have another look tomorrow night and see if I can find something somewhere that it doesnt like.

Was that the `dep` folder you were nuking in the paste? In my experience with HAOS, it's empty, and the lib is installed in `/usr/local/lib/python3.13/site-packages/. Is there anything in your `deps` directory after a restart? I'd expect not.

```
➜  ~ docker exec -it homeassistant bash
homeassistant:/config# cd /usr/local/lib/python3.13/site-packages/
```

> ```TypeError: api.__init__() got an unexpected keyword argument 'id_token'```

Yeah, that still looks like the old lib

shtrom commented:
Ha, now it loses the auth after restarts once the `id_token` has become invalid. The Python lib auto-refreshes the token, but it's not stored HA-side.

I'll update when I get a chance.

shtrom committed:
Save and restore full token accross refreshes

shtrom committed:
Improve error handling and logging

shtrom committed:
Rename class and attributes for clarity

shtrom committed:
Handle AuroraPlusAuthenticationError

shtrom committed:
Linting in sensor

shtrom committed:
Update configentry in async update method

shtrom commented:
I pushed a new version, to better refresh the auth. I'm not 100% sure it's good yet, so only update if you want to test. If so, it needs an update of the AuroraPlus lib, too.

shtrom committed:
more debug logging tweaks

shtrom committed:
Use full token in update_listener if changed

shtrom committed:
readme: one fewer CAVEAT!

shtrom committed:
coordinator: use address from /current rather than /month

shtrom committed:
api: ensure data has been fetched prior to creating sensors

shtrom committed:
readme: muliple services not supported

shtrom commented:
Ok! I pushed a few more changes that should make the automatic reauth work now. Needs an update of the Python lib, too. So update HA integration, delete Python lib, reboot HAOS.

shtrom committed:
coordinator: get tariffs from previous week

shtrom committed:
readme: maintained for 2025

shtrom committed:
manifest: point AuroraPlus back to upstream and add codeowner

shtrom committed:
lint: ruff check --fix + manual fixes

shtrom commented:
Ok, this has been running without an issue for a few weeks. I'm gonna merge all this!

shtrom commented:
I'm going to leave the branch around on my fork, to avoid breaking installs, but everybody who has been pointing HACS to it should now point it back to using the upstream from https://github.com/LeighCurran/AuroraPlusHA/tree/main
